### PR TITLE
Encode object values of Manatee.Json types by copying during serialization

### PR DESCRIPTION
**Pull request by pekaaw**
This solves the problem raised at [Stack Overflow@2020-09-23](https://stackoverflow.com/questions/64029011/how-to-serialize-manatee-json-jsonvalue).

It might introduce problems with getting correct local ```"$ref"``` references.

I have used another testing suite locally, and written a working test for two type of test classes.
The first utilize an ```object``` property, the other a ```JsonValue``` property.
It also works for ```Manatee.Json.JsonObject``` and ```Manatee.Json.JsonArray```.

### TestClass 1
```csharp
public class TestObject
{
    public object Property { get; set; }
}
```

### TestClass 2
```csharp
public class TestObject
{
    public JsonValue Property { get; set; }
}
```

### Test
```csharp
[TestClass]
public class TestClass
{
    [TestMethod]
    public void Test()
    {
        var testObject = new TestObject
        {
            Property = new JsonObject { ["testKey"] = "testValue" }
        };

        JsonValue expected = new JsonValue(
            new JsonObject
            {
                ["Property"] = new JsonObject { ["testKey"] = "testValue" }
            }
        );

        var serializer = new Manatee.Json.Serialization.JsonSerializer();
        JsonValue actual = serializer.Serialize(testObject);

        var expectedString = expected.ToString();
        /// expectedString	"{\"Property\":{\"testKey\":\"testValue\"}}"	string

        var actualString = actual.ToString();
        /// actualString	"{\"Property\":{\"testKey\":\"testValue\"}}"	string

        Assert.AreEqual(expectedString, actualString);
    }
}
```

Side note: I really appreciate this library 👍

### CODE REVIEW

**Comment by pekaaw** [comment]
I'm about to create a ```JsonValueSerializer``` instead, which I think might be a more robust solution.
It should also keep the ```"$ref"``` references correctly when I'm done 👍

**Comment by pekaaw** [comment]
Alright - the `JsonValueSerializer` is ready for review 👍 
What do you think, @gregsdennis?

I've also extended `Manatee.Json.Internal.ReflectionExtensions.SetMember(...)` to enable de-referencing during deserialization.

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        };
+
+        public bool Handles(SerializationContextBase context)
+        {
+            return HandledTypes.Contains(context.InferredType);
+        }
+
+        public JsonValue Serialize(SerializationContext context)
+        {
+            switch (context.Source)
```
Comment (gregsdennis) [inline Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:26]: why are you switching on this and not just returning `context.Source`?  You need to return a `JsonValue` and you're getting a `JsonValue`...

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                context.Push(inferredValue.GetType(), typeof(JsonValue), i.ToString(), inferredValue);
+                var json = context.RootSerializer.Serialize(context);
+                context.Pop();
+                return json;
+            }).ToList().ToJson();
+        }
+
+        public object Deserialize(DeserializationContext context)
+        {
+            switch (context.LocalValue.Type)
```
Comment (gregsdennis) [inline Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:78]: Here you need to check the type that the serializer is requesting over the type that you're given.

File: Manatee.Json/Internal/ReflectionExtensions.cs [hunk 3]
```diff
@@ -87,8 +87,54 @@ public static void SetMember(this object? obj, JsonPointer pointer, object? valu
 				if (int.TryParse(segment, out var index))
 				{
 					member = ReflectionCache.GetMembers(obj.GetType(), PropertySelectionStrategy.ReadWriteOnly, false)
-						.Single(m => m.MemberInfo is PropertyInfo pi && pi.GetIndexParameters().Length == 1);
-					if (member == null) return;
+						.SingleOrDefault(m => m.MemberInfo is PropertyInfo pi && pi.GetIndexParameters().Length == 1);
+					if (member == null)
+					{
+						var jsonArray = (obj as JsonValue)?.Array ?? obj as JsonArray;
```
Comment (gregsdennis) [inline Manatee.Json/Internal/ReflectionExtensions.cs:93]: i don't follow what you're trying to do here.

**Review by gregsdennis — CHANGES_REQUESTED** [review]
please add some testing around these changes.

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs (hunk 1 above)
Comment (pekaaw) [reply Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:26]: I understand it such that what comes in can be any of the three types the serializer can handle: `JsonValue`, `JsonArray` and `JsonObject`. For json literals, the JsonValue are returned, but for `JsonArray` and `JsonObject` I serialize each item/property. The reason for this is to maintain `"$ref"` functionality, just like the other serializers does.

I think if `"$ref"` is not important for `JsonValue`, `JsonArray` nor `JsonObject`, we could just return `context.Source` like you propose?

File: Manatee.Json/Internal/ReflectionExtensions.cs (hunk 3 above)
Comment (pekaaw) [reply Manatee.Json/Internal/ReflectionExtensions.cs:93]: Upon deserialization, it seems like the process tries to replace "$ref" references with the corresponding content.
Here I try to add functionality for resolving "$ref" references on `JsonArray` types (during deserialization).
Without `.SingleOrDefault(...)`, an exception was thrown and the subsequent `if (member == null) return;` would never hit.

Further I try to extract a `JsonArray` instance from `obj`. If it is not null, there is a JsonArray, and values could be added, not by object member, but by indices in the `JsonArray`.

The same goes for `JsonObject` down below, when the segment is not a number, but rather a member name. If the member is `null`, but `obj` is a `JsonObject`, we could proceed with traversing values in the `obj` `JsonObject`.

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs (hunk 2 above)
Comment (pekaaw) [reply Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:78]: I'm not sure if I follow. Is it not such that when deserializing from any to `JsonValue`, `context.RequestedType` is always `JsonValue`?

At least, when I stepped through the code, the requested type was always found to be `JsonValue`.

I could for instance add:
```
if (context.RequestedType != typeof(JsonValue))
{
    throw new NotImplementedException();
}
```

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs (hunk 1 above)
Comment (gregsdennis) [reply Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:26]: If you're deserializing to the JSON content, object references are irrelevant.

File: Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs (hunk 2 above)
Comment (gregsdennis) [reply Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs:78]: Yeah, saying that is what I'm looking for.  The purpose is to deserialize into what the actual model is asking for, not what the serializer _thinks_ should be there.

Think about the case where the property is `JsonArray`.  If you give it a `JsonValue`, it'll fail.

File: Manatee.Json/Internal/ReflectionExtensions.cs (hunk 3 above)
Comment (gregsdennis) [reply Manatee.Json/Internal/ReflectionExtensions.cs:93]: Yeah, same here: we're not concerned about object references inside pure JSON content.

**Comment by gregsdennis** [comment]
@pekaaw if you can make these updates, I'd like to pull them into the next version.

**Comment by pekaaw** [comment]
Thanks. I'll try to get it done during the week. I found a [solution I could use at work](https://stackoverflow.com/a/64276193/3032543), so I'll probably do this in my spare time.

**Comment by gregsdennis** [comment]
Closed to to inactivity.

**Pull request closed without merging**